ddsrecorder_participants/src/cpp/recorder/mcap/McapHandler.cpp

```diff
@@ -119,204 +108,178 @@ McapHandler::~McapHandler()
+    }
 
-        // Create schema and add it to writer and to schemas map
-        std::string encoding = configuration_.ros2_types ? "ros2msg" : "omgidl";
-        mcap::Schema new_schema(configuration_.ros2_types ? utils::demangle_if_ros_type(dynamic_type->get_name()) :
-                dynamic_type->get_name(), encoding, schema_text);
+    mcap::Schema new_schema(name, encoding, data);
 
-        try
-        {
-            // Check if there is enough space on disk to write the schema
-            mcap_size_tracker_.schema_to_write(new_schema);
-        }
-        catch (const std::overflow_error& e)
-        {
-            on_mcap_full_(e, [&]()
-                    {
-                        mcap_size_tracker_.schema_to_write(new_schema);
-                    });
-        }
+    // Add schema to writer and to schemas map
+    logInfo(DDSRECORDER_MCAP_HANDLER, "\nAdding schema with name " << type_name << " :\n" << data << "\n");
 
-        // WARNING: passing as non-const to MCAP library
-        mcap_writer_.addSchema(new_schema);
-        mcap_size_tracker_.schema_written(new_schema);
+    mcap_writer_.write(new_schema);
 
-        logInfo(DDSRECORDER_MCAP_HANDLER, "Schema created: " << new_schema.name << ".");
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Schema created: " << new_schema.name << ".");
 
-        auto it = schemas_.find(type_name);
-        if (it != schemas_.end())
-        {
-            try
-            {
-                // Update channels previously created with blank schema
-                update_channels_nts_(it->second.id, new_schema.id);
-            }
-            catch (const std::overflow_error& e)
-            {
-                // Do not retry if channel update fails (channels are created on the fly when adding data)
-                on_mcap_full_(e);
-            }
-        }
-        schemas_[type_name] = std::move(new_schema);
-        received_types_.insert(type_name);
+    auto it = schemas_.find(type_name);
+    if (it != schemas_.end())
+    {
+        // Update channels previously created with blank schema
+        update_channels_nts_(it->second.id, new_schema.id);
+    }
+    schemas_[type_name] = std::move(new_schema);
+    received_types_.insert(type_name);
 
-        // Every time a dynamic type is added the attachment is newly calculated
-        add_dynamic_type_(type_name);
+    // Every time a dynamic type is added the attachment is newly calculated
+    store_dynamic_type_(type_name, dynamic_types_);
 
-        // Check if there are any pending samples for this new schema. If so, add them.
-        if ((pending_samples_.find(type_name) != pending_samples_.end()) ||
-                (state_ == McapHandlerStateCode::PAUSED &&
-                (pending_samples_paused_.find(type_name) != pending_samples_paused_.end())))
-        {
-            add_pending_samples_nts_(type_name);
-        }
-    }
-    catch (const std::overflow_error& e)
+    if (configuration_.record_types)
```

Comment:
I just realized we should check if the recorded dynamic types are going to be stored in the MCAP file before calculating the attachment as every time we discover a new type we have to serialize it (not optimal if ``record_types=false``).